Remove unnecessary new version from segment descriptor

deepthidevaki opened the pull request:
## Description

Remove keeping track of descriptor length for each version. Instead the length is determined while decoding sbe. As a consequence, we cannot check and fail early before the descriptor is read if the file size is not enough to store the descriptor. 

Since we rely on sbe to determine length, we can also remove version 3 which was introduced recently when new fields were added to the descriptor. 

In addition, this PR also refactors SegmentDescriptor and split the reading from buffer and decoding out of it.

Note:- This is backward **incompatible** with the previous SNAPSHOT and 8.3.0-alpha3, because we removed version 3. But it is still backward compatible with previous minor (8.2.x). 

## Related issues

related #13240 But we decided not use fixed length

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

deepthidevaki commented:
Note: Since the previous version was included in alpha3, this PR breaks compatibility with alpha3. But it is still backward compatible with 8.2.x.

npepinpe commented on journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java:
```diff
@@ -151,27 +152,11 @@ private SegmentDescriptor readDescriptor(final Path file) {
-      buffer.flip();
-      return new SegmentDescriptor(buffer);
+      // We don't know the length of segment descriptor, so we let
+      final var buffer = channel.map(MapMode.READ_ONLY, 0, fileSize);
+      final var descriptor = new SegmentDescriptorReader().readFrom(buffer);
+      IoUtil.unmap(buffer);
```
:upside_down_face: The overhead of mapping/unmapping just to read the descriptor seems unnecessary. As it's not really in the hot path it's probably fine, but would be nice if we got rid of that. Maybe we just map it once initially for the length of the file, then only remap if the descriptor says the segment is actually longer (though unlikely, since we pre-allocate _before_ we write the descriptor)?

deepthidevaki replied:
> Maybe we just map it once initially for the length of the file

Do you mean to re-use the mapping for the Segment?

npepinpe replied:
Yeah, pretty much, so we only map it once in most cases. It's very minor though, can be skipped if it turns out it's a lot of refactoring

npepinpe reviewed (approved):
:rocket: 

This will likely break zdb again though :smile:

deepthidevaki commented on journal/src/main/java/io/camunda/zeebe/journal/file/SegmentDescriptor.java:
```diff
@@ -426,7 +309,169 @@ Builder withMaxSegmentSize(final int maxSegmentSize) {
+    }
+  }
+
+  static final class SegmentDescriptorReader {
```
Agree. We can remove version 1. I will update the PR. 

> the general concept of reading multiple versions to create the same model lends itself very well to a strategy pattern, 

Thanks for the suggestion. Let me try it out.

deepthidevaki commented:
@npepinpe I have applied your comments:
- Moved the reader to own file
- Reuse mapped buffer for the segment and reading the descriptor
- removed support for version 1. 

I didn't apply the strategy pattern as you suggested. There is only one version now, so it doesn't matter much. But it will be valuable when we add new versions later.

Feel free to take another look. Otherwise I will merge this on Monday.

npepinpe commented:
Go for it, no need for another review. And yeah, doesn't make sense to introduce strategies if we only have one version now :smile:

deepthidevaki commented:
bors merge
